recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+                    )
+                (backprop_loss / self.grad_accumulation_factor).backward()
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    self.optimizer_step += 1
+
+        if should_step:
+            if isinstance(
+                self.modules.model, DistributedDataParallel
+            ):
+                w2v_model = self.modules.model.module
+            else:
+                w2v_model = self.modules.model
+            if w2v_model.quantiser is not None:
+                w2v_model.quantiser.update_temp(
+                    self.optimizer_step
+                )
+
+            if (
+                hasattr(self.hparams, "log_interval")
+                and self.optimizer_step % self.hparams.log_interval
+                == 0
+            ):
+
+                log_dct = {
+                    k: (
+                        v.item()
+                        if isinstance(v, torch.Tensor)
+                        else v
+                    )
+                    for k, v in objectives.items()
+                }
+                current_lr = self.optimizer.param_groups[0]["lr"]
+                log_dct["lr"] = current_lr
+                log_dct["avg_loss"] = self.avg_train_loss
+
+                if hasattr(self, "time_last_log"):
+                    run_time_since_last_log = (
+                        time.time() - self.time_last_log
+                    )
+                    log_dct["stats/run_time"] = run_time_since_last_log
+                self.time_last_log = time.time()
+
+                log_str = f"Update: {self.optimizer_step} - Objectives: {log_dct}"
+                logger.info(log_str)
+
+                if self.hparams.use_wandb:
+                    run_on_main(
+                        wandb.log,
+                        kwargs={
+                            "data": log_dct,
+                            "step": self.optimizer_step,
+                        },
+                    )
+
+            self.hparams.lr_scheduler(
+                self.optimizer, self.optimizer_step
+            )
+
+        return objectives["loss"].detach()
+
+    def evaluate_batch(self, batch, stage):
+        out = self.compute_forward(batch, stage=stage)
+        objectives = self.compute_objectives(out, batch, stage=stage)
+        return objectives["accuracy"].cpu()
+
+    def on_stage_end(self, stage, stage_loss, epoch=None):
+        if stage == sb.Stage.VALID:
+            logger.info(
+                f"Update: {self.optimizer_step} - valid_accuracy: {stage_loss:.3f}"
+            )
+            if self.hparams.use_wandb:
+                wandb.log(
+                    {
+                        "valid_acuracy": stage_loss,
+                    },
+                    step=self.optimizer_step,
+                )
+
+    def update_average(self, loss, avg_loss):
```

Comment:
what is this function necessary for ?

Reply:
The default `update_average` [link](https://github.com/speechbrain/speechbrain/blob/a9b373a9fef3fdfc4e3b4a41478f7b283866186f/speechbrain/core.py#L1239) has two issues for me:

1. it uses `self.step` which is reset every epoch which seems wrong to me (why would you want to have such a reset after every epoch?)
2. it's not clear to me why/how the implementation works

so I used a simple exponential moving average